Add function to autosave the project

pankern opened the pull request:
The autosaves are implemented here in a rolling backup fashion, such that the 10 most recent autosaves are kept. The autosaves don't overwrite the main save file. The changes include modifying the `incSaveProject` function in order to avoid adding the autosave files to the list of previous projects.

I thought it would be good to get this PR going with what I have now so I could incorporate feedback sooner rather than later.

Things not included in this PR that will likely still be needed:

- Ability to turn the autosave feature on/off.
- Ability to configure how many backups are kept.
- Ability to configure how often to trigger the autosave.
- Improve the performance of the filesave so that the program doesn't freeze as long during autosave. This might not be as important but I did notice it during testing.

LunaTheFoxgirl commented on source/creator/package.d:
```diff
@@ -119,6 +123,77 @@ string incProjectPath() {
+    //We'll add the extension back later when we need it.
+    path = path.stripExtension;
+
+    string pathBaseName = path.baseName;
```
backups should be saved in the app data directory, see the creator.core.path module.
We can't trust that the base name will be writable on all platforms.

LunaTheFoxgirl commented on source/creator/package.d:
```diff
@@ -216,12 +293,14 @@ void incOpenProject(string path) {
-        incAddPrevProject(finalPath);
+        if (!autosave) {
+            currProjectPath = path;
+            incAddPrevProject(finalPath);
```
We probably want a store buffer of auto-saves that can be listed as well.

Maybe we should leave a inochi-creator.lock file in the app data dir when a project is loaded or created so that we can display a popup asking whether to load the last backup? idea would be to delete the lock file when the app closes normally.

pankern replied:
Buffer of recent autosaves: Sure that sounds good. Since the real filenames will end up long, I am considering having the user-facing portion show the datetime made (in short format) and the project name that the autosave is for?

Lock file: I like this approach for easy crash recovery. Creating the lock file can be done on the first autosave of that app session, since before that there won't be anything to restore from (or at least nothing more recent that what we loaded from).

LunaTheFoxgirl replied:
There is one thing I worry about which is what if you start multiple Inochi Creator instances?
I'm not sure we can feasibly in a cross platform way check whether Inochi Creator is already running...

pankern replied:
It might be possible to work around this by having the lockfiles be specific to each project file (storing it with the autosaves). I suppose in the worst case scenario someone could open the same project in multiple Inochi Creator instances, such as if their workspace is disorganized and they didn't realize it was already open.

We could also compare the timestamps of autosaves with the timestamp of the main project file as a way of seeing if there is work to recover, if that's something we think can be relied on.

LunaTheFoxgirl reviewed (changes requested):
There's a few changes I'd like to see with what's currently submitted and some ideas on where to go from there.

LunaTheFoxgirl commented on source/creator/package.d:
```diff
@@ -119,6 +123,77 @@ string incProjectPath() {
+    Create a backup save path string.
+*/
+string bakFilename(string path, int bakNum) {
+    string bakName = format("%.2s", bakNum);
```
Maybe a good idea to have the date and time the file was autosaved in the name, or at the very least in an internal list?

pankern replied:
I will put the date and time in the filename instead of the incremented recursion number. For choosing which file to remove once the backup limit is reached, it should then be possible to just remove the first one alphabetically. I feel like I would trust that more than trusting the OS didn't muddy with file timestamps, unless I'm being overly paranoid.

pankern commented:
I have addressed most of the comments. One thing I haven't figured out yet is how to elegantly load any of the autosaves displayed in the menu. See the TODO I added to `mainmenu.d`.
I also still need to implement the user settings for the autosaves, but I think that will be quick and easy now that I have a better handle on more of the code.

pankern commented:
I have completed the code for properly opening autosaves from the recents menu as well as implementing the user settings panel for controlling the autosave behavior. This is now ready for additional review.

pankern commented:
Ok, I will rebase this feature on the latest from main.
